### PR TITLE
Run DNS lookup once and validate it before starting

### PR DESCRIPTION
**Pull request by chadell**
* IPv4 and IPv6 resolution
* static array selection, review

### CODE REVIEW

**Comment by chadell** [comment]
Lookup support IPv4 and IPv6 resolution: https://golang.org/src/net/lookup.go?s=5208:5248#L157

**Comment by chadell** [comment]
Regarding the static access to the first element of the array `addrs[0]`. If the resolution has worked well it will be always a first entry. Another way, it could be a random selection in addrs[], but I don't see too much value...

**Comment by chadell** [comment]
@dcaba have a look at my two comments ;)

File: cmd/tcpgoon.go [hunk 1]
```diff
@@ -63,6 +65,13 @@ func validateRequiredArgs(params *tcpgoonParams, args []string) error {
 		return errors.New("Number of required parameters doesn't match")
 	}
 	params.hostPtr = args[0]
+
+	addrs, err := net.LookupIP(params.hostPtr)
```
Comment (dcaba) [inline cmd/tcpgoon.go]: what if we already got an IP address?

File: cmd/tcpgoon.go [hunk 2]
```diff
@@ -63,6 +65,13 @@ func validateRequiredArgs(params *tcpgoonParams, args []string) error {
...
+	if err != nil {
+		return errors.New("Domain name not resolvable")
+	}
+	params.hostPtr = addrs[0].String()
```
Comment (dcaba) [inline cmd/tcpgoon.go]: is there a corner case where you may not get an err but the slice is empty? Should we check then the slice size before checking the first element?

Comment (dcaba) [inline cmd/tcpgoon.go]: I don't think rewriting the original variable is a good idea. Probably it makes sense to store both, for reporting purposes. Please, consider:
* to inform, at least, in the debug output of the IP address you pick to run the tests against
* what do you think makes more sense: to print in the actual report the IP address we are using, the dns or both (something like `dns(ip)`)?

File: cmd/tcpgoon.go (hunk 1 above)
Comment (chadell) [reply cmd/tcpgoon.go]: It's solved by the LookupIP  (https://golang.org/src/net/lookup.go#L177)

File: cmd/tcpgoon.go (hunk 2 above)
Comment (chadell) [reply cmd/tcpgoon.go]: I don't think so, but we can add this check

Comment (chadell) [reply cmd/tcpgoon.go]: done

File: cmd/tcpgoon.go [hunk 3]
```diff
@@ -56,7 +59,15 @@ func validateRequiredArgs(params *tcpgoonParams, args []string) error {
 	if len(args) != 2 {
 		return errors.New("Number of required parameters doesn't match")
 	}
-	params.hostPtr = args[0]
+	params.hostName = args[0]
```
Comment (dcaba) [inline cmd/tcpgoon.go]: as hostName may not be a hostname (but an IP address), I'd consider this as "target"

Comment (dcaba) [reply cmd/tcpgoon.go]: and i would put, in any case, hostname and not hostName, wouldnt you?

File: cmd/tcpgoon.go [hunk 4]
```diff
@@ -15,6 +16,7 @@ import (
 
 type tcpgoonParams struct {
 	hostPtr              string
```
Comment (dcaba) [inline cmd/tcpgoon.go]: do the "Ptr" names make sense any more? If they are just strings/ints, and not pointers to strings/ints (as happened with the args mgmt methods in the standard library), I'd rename

File: cmd/tcpgoon.go [hunk 5]
```diff
@@ -56,7 +59,15 @@ func validateRequiredArgs(params *tcpgoonParams, args []string) error {
...
 		return errors.New("Number of required parameters doesn't match")
 	}
-	params.hostPtr = args[0]
+	params.hostName = args[0]
+	addrs, err := net.LookupIP(params.hostName)
+	if err != nil || len(addrs) == 0 {
+		return errors.New("Domain name not resolvable")
+	}
+
+	params.hostPtr = addrs[0].String()
```
Comment (dcaba) [inline cmd/tcpgoon.go]: hostPtr is an IP address, not a hostPtr, right?

**Comment by dcaba** [comment]
See my commits, as it addresses some of my comments (remember to pull them in your local git repo!). Then only variable naming matters should require a look...

File: _script/cibuild-docker [hunk 6]
```diff
@@ -71,7 +77,7 @@ function docker_run_cmdline_generator() {
             $docker_deploy_flags \
             --privileged \
             --rm $test_docker_image \
-            /bin/bash --login -c _script/$script_name"
+            /bin/bash --login -c \"_script/$script_name $dry_run \""
```
Comment (chadell) [inline _script/cibuild-docker]: `\""`?

**Pull request merged**